### PR TITLE
FriendshipsActionRequest

### PR DESCRIPTION
**Pull request by noirtier-villefort**
FriendshipsActionResponse
FriendshipsActionTest

Let user follow and unfollow

### CODE REVIEW

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import io.github.taz03.jia.TestProxy;
+import io.github.taz03.jia.requests.friendships.FriendshipsActionRequest;
+import io.github.taz03.jia.responses.friendships.FriendshipsActionResponse;
+import org.junit.jupiter.api.Test;
+
+public class FriendshipsActionTest {
+    @Test
+    public void unfollowPkTest() throws Exception{
+        long pk = 3333; // your pk here
```
Comment (Taz03) [inline lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:13]: use pk from proxy client

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        FriendshipsActionResponse response = TestProxy.client.sendRequest(new FriendshipsActionRequest(pk, FriendshipsActionRequest.FriendshipsAction.DESTROY)).get();
```
Comment (Taz03) [inline lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:14]: Create request and response separately

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import io.github.taz03.jia.TestProxy;
+import io.github.taz03.jia.requests.friendships.FriendshipsActionRequest;
+import io.github.taz03.jia.responses.friendships.FriendshipsActionResponse;
+import org.junit.jupiter.api.Test;
+
+public class FriendshipsActionTest {
+    @Test
+    public void unfollowPkTest() throws Exception{
```
Comment (Taz03) [inline lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:12]: first do the follow test and then unfollow the profile, for this test use instagram's official account who's pk is `25025320`

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package io.github.taz03.jia.responses.friendships;
+
+import io.github.taz03.jia.responses.InstagramResponse;
+import io.github.taz03.jia.responses.models.friendship.Status;
+
+public class FriendshipsActionResponse extends InstagramResponse {
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:6]: see other response classes and follow

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    private Status friendship_status;
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:7]: use java naming convention

**Review by Taz03 — CHANGES_REQUESTED** [review]
fix formatting

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java (hunk 3 above)
Comment (noirtier-villefort) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:12]: do follow inside unfollow or move follow method to top of the code? 
use instagram`s official account pk or [pk from proxy client](https://github.com/Taz03/JIA/pull/37#discussion_r1314276763) ?

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java (hunk 2 above)
Comment (noirtier-villefort) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:14]: fixed

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java (hunk 3 above)
Comment (Taz03) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:12]: lets say person A following person B so, A should be us (proxy client pk) and B should be instagram's official account (the pk I mentioned)

Comment (Taz03) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:12]: just noticed A is not in the equation here so yea use instagram's official accounts pk

Comment (noirtier-villefort) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java]: maybe make pk as global variable for 2 tests? No sense to use local

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java (hunk 4 above)
Comment (noirtier-villefort) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:10]: I set annotations. Do you mean it?

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java (hunk 5 above)
Comment (noirtier-villefort) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java]: renamed and setted @JsonProperty annotation

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java (hunk 4 above)
Comment (Taz03) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:10]: also add getter

**Review by Taz03 — CHANGES_REQUESTED** [review]
docs and fomatting

**Comment by noirtier-villefort** [comment]
oh.. something wrong

**Comment by Taz03** [comment]
> is it ok that [ebdd3e7](https://github.com/Taz03/JIA/pull/37/commits/ebdd3e7b9ddfdd27769f5a4b52c33f2ffe7753a9) [7ddd4eb](https://github.com/Taz03/JIA/pull/37/commits/7ddd4eb6047dc6e8ff0011628b0ba7b8b79a0792), [27cf26e](https://github.com/Taz03/JIA/pull/37/commits/27cf26e9f5b6abbbc914d922cdb17f19df199524) here?

No you should fetch and merge or fetch and rebase, learn git

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import io.github.taz03.jia.TestProxy;
+import io.github.taz03.jia.requests.friendships.FriendshipsActionRequest;
+import io.github.taz03.jia.responses.friendships.FriendshipsActionResponse;
+import org.junit.jupiter.api.Test;
+
+public class FriendshipsActionTest {
+    private final long pk = 25025320;
```
Comment (Taz03) [inline lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:11]: static, name it `INSTAGRAM_PK` and explain in a comment maybe also give the link to profile

File: lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+     * Creates an Instagram action request
+
+     * @param pk user's pk to apply the action on him
+     * @param action action of {@link FriendshipsAction} to follow or unfollow
+     */
+    public FriendshipsActionRequest(long pk, FriendshipsAction action){
+        super(FriendshipsActionResponse.class, String.format("/api/v1/friendships/%s/%s/", action.name().toLowerCase(), pk));
+    }
+
+    public enum FriendshipsAction{
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java:21]: write a small doc
```
CREATE = follow
DESTROY = unfollow
```

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import io.github.taz03.jia.responses.InstagramResponse;
+import io.github.taz03.jia.responses.models.friendship.Status;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class FriendshipsActionResponse extends InstagramResponse {
+   @JsonProperty("friendship_status")
+    private Status friendshipStatus;
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:12]: i see that it needs more than just `Status` and also for the recently added show friendships request I'll be adding an `ExtendedStatus` model

Comment (Taz03) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:12]: Done. U can now use `ExtendedStatus`

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:1]: also add toString, we forgot it last time  😅

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+public class FriendshipsActionRequest extends InstagramPostRequest<FriendshipsActionResponse> {
+
+    /**
+     * Creates an Instagram action request
+
+     * @param pk user's pk to apply the action on him
+     * @param action action of {@link FriendshipsAction} to follow or unfollow
+     */
+    public FriendshipsActionRequest(long pk, FriendshipsAction action){
+        super(FriendshipsActionResponse.class, String.format("/api/v1/friendships/%s/%s/", action.name().toLowerCase(), pk));
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java:18]:
```suggestion
        super(FriendshipsActionResponse.class, "/api/v1/friendships/%s/%d/".formatted(action.name().toLowerCase(), pk));
```

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java (hunk 6 above)
Comment (noirtier-villefort) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java:11]: i see you already change it at main commits. I do git pull. Is it right?

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java (hunk 8 above)
Comment (noirtier-villefort) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:12]: i changed Status to ExtendedStatus. Should i change JsonProperty value?

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java
Comment (noirtier-villefort) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:1]: added

File: lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java (hunk 7 above)
Comment (noirtier-villefort) [reply lib/src/main/java/io/github/taz03/jia/requests/friendships/FriendshipsActionRequest.java:21]: added

**Comment by noirtier-villefort** [comment]
force-push mistake

File: lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java (hunk 6 above)
Comment (Taz03) [reply lib/src/integrationTest/java/io/github/taz03/jia/friendships/FriendshipsActionTest.java]: This pr is not conflicting  so if u want  u can continue, but if u want to rebase first checkout `main` then do fetch and pull after that checkout this branch again and do a merge or rebase. But that is not required

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+   @JsonProperty("friendship_status")
+    private ExtendedStatus friendshipStatus;
+
+    public ExtendedStatus getFriendshipStatus() {
+        return friendshipStatus;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString();
```
Comment (Taz03) [inline lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java:20]: write a proper toString method see other classes

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java (hunk 8 above)
Comment (Taz03) [reply lib/src/main/java/io/github/taz03/jia/responses/friendships/FriendshipsActionResponse.java]: no, u r still parsing the same json value just as another format, it'll  stay the same

**Review by Taz03 — APPROVED** [review]

**Pull request merged**